README.md
# TipDialog
> 1. Please use 3.0.0 or above version, because earlier versions have serious performance issues
> 2. if example no working, please google questions about Flutter upgrade AndroidX,
> [AndroidX Migration - Flutterblock](https://flutter.dev/docs/development/androidx-migration)
> 3. if run your app, it shows `Error: Cannot run with sound null safety`, please upgrade version to 4.0.0
> [Reference](https://dart.dev/null-safety/migration-guide)

[中文版](https://github.com/inRush/TipDialog/blob/master/README.zh-CN.md)

A Flutter Tip Dialog

| Loading Type Dialog | Success Tye Dialog | Fail Type Dialog |
| --- | --- | --- |
|  ![Loading](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/loading.jpeg) | ![Success](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/success.jpeg)| ![Fail](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/failed.jpeg) |


| Info Type Dialog | Only Icon Dialog | Onl Text Dialog  |
| --- | --- | --- |
| ![Info](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/info.jpeg)| ![OnlyIcon](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/only-icon.jpeg)| ![OnlyText](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/only-text.jpeg)|


| Custom Icon Dialog | Custom Body Dialog |
| --- | --- |
| ![CustomIcon](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/custom-icon.jpeg)| ![CustomBody](https://raw.githubusercontent.com/inRush/Resources/master/images/tip-dialog/custom-body.jpeg)|


## 1. Depend on it
Add this to your package's pubspec.yaml file:

``` dart
dependencies:
  tip_dialog: ^4.0.0
```
## 2. Install it
You can install packages from the command line:
with Flutter:

```
$ flutter packages get
```
## 3. Import it
Now in your Dart code, you can use:

```dart
import 'package:tip_dialog/tip_dialog.dart';
```
## 4. Use
#### Available attributes

```
/// [TipDialogContainer]
@required this.child,
/// automatically disappear time
this.duration: const Duration(seconds: 3),
/// mask layer alpha
this.maskAlpha: 0.3
// outside touchable, default false
// if true and onOutsideTouch not set, when use TipDialogHelper.loading(), you can touch mask layer to dismiss
// if true and onOutsideTouch set, when touching the mask layer, onOutsideTouch will be called.
this.outsideTouchable: false
// outside touch callback {@link OutsideTouchCallback}
this.onOutsideTouch
```

#### Global Use
```dart
/// Use [TipDialogContainer] globally
/// This widget can be globally supported
void main() => runApp(new MyApp());
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'TipDialog Demo',
      theme: new ThemeData(),
      home: Stack(
        children: <Widget>[
          MyHomePage(title: 'TipDialog Demo Home Page'),
          // add [TipDialogContainer] widget here
          TipDialogContainer(duration: const Duration(seconds: 2))
        ],
      ),
    );
  }
}
/// use TipDialogHelper to show or dismiss tip
  Widget build(BuildContext context) {
    return new Scaffold(
        appBar: new AppBar(
          title: new Text(widget.title),
          elevation: 0.5,
        ),
        body: new ListView(children: <Widget>[
          _buildItem("Loading Type Tip Dialog", () async {
            TipDialogHelper.loading("Loading");
            await new Future.delayed(new Duration(seconds: 5));
            TipDialogHelper.dismiss();
          }),
          new Divider(),
          _buildItem("Success Type Tip Dialog", () async {
            TipDialogHelper.success("Loaded Successfully");
          }),
          new Divider(),
          _buildItem("Only text Tip Dialog", () async {
            TipDialogHelper.show(new TipDialog(
              type: TipDialogType.NOTHING,
              tip: "Do Not Repeat",
            ));
          }),
          new Divider(),
          _buildItem("Custom Icon Tip Dialog", () async {
            TipDialogHelper.show(new TipDialog.customIcon(
              icon: new Icon(
                Icons.file_download,
                color: Colors.white,
                size: 30.0,
                textDirection: TextDirection.ltr,
              ),
              tip: "Download",
            ));
          }),
          new Divider(),
          _buildItem("Custom Body Tip Dialog", () async {
            TipDialogHelper.show(new TipDialog.builder(
              bodyBuilder: (context) {
                return new Container(
                  width: 120.0,
                  height: 90.0,
                  alignment: Alignment.center,
                  child: new Text(
                    "Custom",
                    style: new TextStyle(color: Colors.white),
                    /// if TipDialogContainer are outside of MaterialApp,
                    /// here is a must to set
                    textDirection: TextDirection.ltr,
                  ),
                );
              },
              color: Colors.blue.withAlpha(150),
            ));
          }),
          new Divider(),
        ]));
  }
```

>Use a custom widget when using [TipDialogContainer] globally, there may be appear some unexpected errors.
>such as Text or Icon, will appear similar to the following error.

***No Directionality widget found.***

>Just set TextDirection just fine. See the custom Widget in the example for details.

## 5. Default Dialog Type
```
enum TipDialogType { NOTHING, LOADING, SUCCESS, FAIL, INFO }

NOTHING: no icon
LOADING: have a loading icon
SUCCESS: have a success icon
FAIL: have a fail icon
INFO: have a info icon
CUSTOM: custom tip dialog type, just a sign, do nothing
```
## 6. TipDialogHelper Method

```dart
/// tipDialog: Need to display the widget
///
/// isAutoDismiss: decide whether to disappear automatically, default is true
/// if true, the dialog will not automatically disappear
/// otherwise, the dialog will automatically disappear after the [Duration] set by [TipDialogContainer]
void show(Widget tipDialog, {bool isAutoDismiss: true});

/// dismiss dialog
void dismiss();

/// show info dialog
void info(String tip);
/// show fail dialog
void fail(String errMsg);
/// show success dialog
void success(String success);
/// show loading dialog
void loading(String loadingTip);
```

>See the example directory for more details.


## 7. Change log

### [4.0.0]

* migrating to null safety

### [3.1.0]

* add outside touchable features

### [3.0.0]

* fix performance issues
* easier way to call

### [2.1.0]

* add TipDialogHelper and deprecated TipDialogConnector

### [2.0.1]

* fix dismiss bug
* upgrade android build gradle to 5.1.1

### [2.0.0] 

* set default auto dismiss duration as 2 seconds
* delete [TipDialogContainer] Partial parameters
    -- show
    -- outSideTouchable
    -- defaultTip
    -- defaultType
* change show method parameter isLoading to isAutoDismiss
* force display mask layer

### [1.1.2] - (MODIFY)

* fix infinite call dismiss bug
* fixed an issue where setting the isLoading value is invalid

### [1.1.1] - (MODIFY)

* fix bugs that occur when using globally

### [1.1.0] - (FUNCTION CHANGE)

* add tip dialog global support

### [1.0.1] - (MODIFY).

* fix loading view version bug.
* set default loading duration

### [1.0.0] - first release.

* add release.


CHANGELOG.md
## [1.0.0] - first release.

* add release.

## [1.0.1] - (MODIFY).

* fix loading view version bug.
* set default loading duration

## [1.1.0] - (FUNCTION CHANGE)

* add tip dialog global support

## [1.1.1] - (MODIFY)

* fix bugs that occur when using globally

## [1.1.2] - (MODIFY)

* fix infinite call dismiss bug
* fixed an issue where setting the isLoading value is invalid

## [2.0.0] 

* set default dismiss duration as 2 seconds
* delete TipDialogContainer parameters
    -- show
    -- outSideTouchable
    -- defaultTip
    -- defaultType
* change show method parameter isLoading to isAutoDismiss
* force display mask layer

## [2.0.1]

* fix dismiss bug
* upgrade android build gradle to 5.1.1

## [2.1.0]

* add TipDialogHelper and deprecated TipDialogConnector

## [3.0.0]

* fix performance issues
* easier way to call

## [3.1.0]

* add outside touchable features

### [4.0.0]

* migrating to null safety

lib/tip_dialog.dart
library tip_dialog;

import 'dart:async';
import 'package:flutter/material.dart';
import 'package:loading_view/loading_view.dart';

enum TipDialogType { NOTHING, LOADING, SUCCESS, FAIL, INFO, CUSTOM }

class TipDialogIcon extends StatelessWidget {
  TipDialogIcon(this.type, {this.color: Colors.white});

  final TipDialogType type;
  final Color color;

  @override
  Widget build(BuildContext context) {
    switch (type) {
      case TipDialogType.SUCCESS:
        return new ImageIcon(
          AssetImage("packages/tip_dialog/images/icon_notify_done.png"),
          size: 35.0,
          color: color,
        );
      case TipDialogType.FAIL:
        return new ImageIcon(
          AssetImage("packages/tip_dialog/images/icon_notify_error.png"),
          size: 35.0,
          color: color,
        );
      case TipDialogType.INFO:
        return new ImageIcon(
          AssetImage("packages/tip_dialog/images/icon_notify_info.png"),
          size: 35.0,
          color: color,
        );
      case TipDialogType.LOADING:
        return new LoadingView(
          35.0,
          color: color,
        );
      default:
        throw new Exception(
            "this type $type is not in TipDialogType: NOTHING, LOADING, SUCCESS, FAIL, INFO");
    }
  }
}

class TipDialog extends StatelessWidget {
  TipDialog({Key? key, this.type: TipDialogType.NOTHING, this.tip})
      : icon = type == TipDialogType.NOTHING ? null : new TipDialogIcon(type),
        _bodyBuilder = null,
        color = const Color(0xbb000000),
        super(key: key);

  TipDialog.customIcon({Key? key, this.icon, this.tip})
      : assert(icon != null || tip != null),
        _bodyBuilder = null,
        type = TipDialogType.CUSTOM,
        color = const Color(0xbb000000),
        super(key: key);

  TipDialog.builder(
      {Key? key,
      WidgetBuilder? bodyBuilder,
      this.color: const Color(0xbb000000)})
      : this._bodyBuilder = bodyBuilder,
        type = TipDialogType.CUSTOM,
        tip = null,
        icon = null,
        super(key: key);

  final String? tip;
  final Widget? icon;
  final WidgetBuilder? _bodyBuilder;
  final Color color;
  final TipDialogType type;

  Widget _buildBody() {
    List<Widget> childs = [];
    if (icon != null) {
      childs.add(new Padding(
        padding: tip == null
            ? const EdgeInsets.all(20.0)
            : const EdgeInsets.fromLTRB(0.0, 8.0, 0.0, 8.0),
        child: icon,
      ));
    }
    if (tip != null) {
      childs.add(new Padding(
        padding: icon == null
            ? const EdgeInsets.fromLTRB(8.0, 0.0, 8.0, 0.0)
            : const EdgeInsets.fromLTRB(8.0, 0.0, 8.0, 8.0),
        child: new Text(
          tip ?? "",
          textAlign: TextAlign.center,
          style: new TextStyle(color: Colors.white, fontSize: 15.0),
          textDirection: TextDirection.ltr,
        ),
      ));
    }
    return new Column(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.center,
      children: childs,
    );
  }

  @override
  Widget build(BuildContext context) {
    return new ClipRRect(
      borderRadius: new BorderRadius.circular(15.0),
      child: new Container(
        constraints: icon == null || tip == null
            ? new BoxConstraints(minHeight: 50.0, minWidth: 100.0)
            : new BoxConstraints(minHeight: 90.0, minWidth: 120.0),
        color: color,
        child: _bodyBuilder == null ? _buildBody() : _bodyBuilder!(context),
      ),
    );
  }
}

typedef OutsideTouchCallback(Widget tipDialog);

class TipDialogContainer extends StatefulWidget {
  TipDialogContainer(
      {this.duration: const Duration(seconds: 2),
      this.maskAlpha: 0.3,
      this.outsideTouchable: false,
      this.onOutsideTouch});

  final Duration duration;
  final double maskAlpha;
  final bool outsideTouchable;

  final OutsideTouchCallback? onOutsideTouch;

  @override
  State<StatefulWidget> createState() {
    TipDialogContainerState state = TipDialogContainerState();
    TipDialogHelper._init(state);
    return state;
  }
}

class TipDialogContainerState extends State<TipDialogContainer>
    with TickerProviderStateMixin {
  Timer? _timer;
  bool _show = false;
  late AnimationController _animationController;
  late Animation<double> _scaleAnimation;
  late VoidCallback _animationListener;
  bool _prepareDismiss = false;
  late Widget _tipDialog;

  /// if true, the dialog will not automatically disappear
  /// otherwise, the dialog will automatically disappear after the [Duration] set by [TipDialogContainer]
  bool _isAutoDismiss = true;

  bool get isShow => _show;

  void dismiss() {
    setState(() {
      if (_prepareDismiss || !_show) {
        return;
      }
      if (_animationController.isAnimating) {
        _show = false;
        _animationController.stop(canceled: true);
      } else {
        _prepareDismiss = true;
        _animationController.reverse();
      }
    });
  }

  /// tipDialog: Need to display the widget
  /// (default uses the dialog set by [TipDialogContainer])
  ///
  /// isLoading: decide whether to disappear automatically
  /// (default uses the value set by [TipDialogContainer],
  /// set type = TipDialogType.LOADING, the value will be true, otherwise will be false.)
  void show(Widget tipDialog, {bool isAutoDismiss: true}) {
    _tipDialog = tipDialog;
    // when tip dialog equal null, isLoading must inherit the origin value
    _isAutoDismiss = isAutoDismiss;
    setState(() {
      _start();
      _show = true;
    });
  }

  @override
  void initState() {
    super.initState();
    _show = false;
    _animationController = new AnimationController(
        value: 0.0, duration: new Duration(milliseconds: 200), vsync: this);
    _animationListener = () {
      if (_animationController.value == 0.0 && _prepareDismiss) {
        setState(() {
          _show = false;
          _prepareDismiss = false;
        });
      }
    };
    _animationController.addListener(_animationListener);
    _scaleAnimation =
        new Tween(begin: 0.95, end: 1.0).animate(_animationController);
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (_show) {
      _start();
    }
  }

  void _start() {
    _animationController.forward(from: 0.0);
    if (_isAutoDismiss) {
      if (_timer != null) {
        _timer?.cancel();
        if (_show) {
          dismiss();
        }
        _timer = null;
      }
      _timer = new Timer(widget.duration, () {
        dismiss();
        _timer = null;
      });
    }
  }

  @override
  void dispose() {
    if (_timer != null) {
      _timer?.cancel();
      if (_show) {
        dismiss();
      }
    }
    _animationController.removeListener(_animationListener);
    _animationController.dispose();
    super.dispose();
  }

  Widget _buildMaskLayer() {
    return new LayoutBuilder(builder: (context, size) {
      return new FadeTransition(
        opacity: _animationController,
        child: new GestureDetector(
          onTap: () {
            if (widget.outsideTouchable) {
              if (widget.onOutsideTouch == null && !_isAutoDismiss) {
                this.dismiss();
              } else if (widget.onOutsideTouch != null) {
                widget.onOutsideTouch!(_tipDialog);
              }
            }
          },
          child: new Container(
            width: size.maxWidth,
            height: size.maxHeight,
            color: Colors.black.withAlpha(widget.maskAlpha * 255 ~/ 1),
          ),
        ),
      );
    });
  }

  @override
  Widget build(BuildContext context) {
    List<Widget> widgets = [];
    if (_show) {
      widgets.add(_buildMaskLayer());
      widgets.add(new ScaleTransition(
        scale: _scaleAnimation,
        child: new FadeTransition(
          opacity: _animationController,
          child: _tipDialog,
        ),
      ));
      return new Scaffold(
        backgroundColor: Colors.transparent,
        body: Stack(
          alignment: Alignment.center,
          children: widgets,
        ),
      );
    }
    return SizedBox.shrink();
  }
}

class TipDialogHelper {
  static late TipDialogContainerState _tipDialog;

  static _init(TipDialogContainerState state) {
    _tipDialog = state;
  }

  static void show(Widget tipDialog, {bool isAutoDismiss: true}) {
    _tipDialog.show(tipDialog, isAutoDismiss: isAutoDismiss);
  }

  static void dismiss() {
    _tipDialog.dismiss();
  }

  static void info(String tip) {
    show(TipDialog(type: TipDialogType.INFO, tip: tip));
  }

  static void fail(String errMsg) {
    show(TipDialog(type: TipDialogType.FAIL, tip: errMsg));
  }

  static void success(String success) {
    show(TipDialog(type: TipDialogType.SUCCESS, tip: success));
  }

  static void loading(String loadingTip) {
    show(TipDialog(type: TipDialogType.LOADING, tip: loadingTip),
        isAutoDismiss: false);
  }
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:tip_dialog/tip_dialog.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'TipDialog Demo',
      theme: new ThemeData(),
      home: Stack(
        children: <Widget>[
          MyHomePage('TipDialog Demo Home Page'),
          TipDialogContainer(
              duration: const Duration(seconds: 2),
              outsideTouchable: true,
              onOutsideTouch: (Widget tipDialog) {
                if (tipDialog is TipDialog &&
                    tipDialog.type == TipDialogType.LOADING) {
                  TipDialogHelper.dismiss();
                }
              })
        ],
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage(this.title,{Key? key}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Widget _buildItem(String name, VoidCallback callback) {
    return new GestureDetector(
      onTap: callback,
      child: new ListTile(
        title: new Text(name),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
        appBar: new AppBar(
          title: new Text(widget.title),
          elevation: 0.5,
        ),
        body: new ListView(children: <Widget>[
          _buildItem("Loading Type Tip Dialog", () async {
            TipDialogHelper.loading("Loading");
            await new Future.delayed(new Duration(seconds: 5));
            TipDialogHelper.dismiss();
          }),
          new Divider(),
          _buildItem("Success Type Tip Dialog", () async {
            TipDialogHelper.success("Loaded Successfully");
          }),
          new Divider(),
          _buildItem("Fail Type Tip Dialog", () async {
            TipDialogHelper.fail("Load Failed");
          }),
          new Divider(),
          _buildItem("Info Type Tip Dialog", () async {
            TipDialogHelper.info("Do Not Repeat");
          }),
          new Divider(),
          _buildItem("Only Icon Tip Dialog", () async {
            TipDialogHelper.show(new TipDialog(
              type: TipDialogType.SUCCESS,
            ));
          }),
          new Divider(),
          _buildItem("Only text Tip Dialog", () async {
            TipDialogHelper.show(new TipDialog(
              type: TipDialogType.NOTHING,
              tip: "Do Not Repeat",
            ));
          }),
          new Divider(),
          _buildItem("Custom Icon Tip Dialog", () async {
            TipDialogHelper.show(new TipDialog.customIcon(
              icon: new Icon(
                Icons.file_download,
                color: Colors.white,
                size: 30.0,
                textDirection: TextDirection.ltr,
              ),
              tip: "Download",
            ));
          }),
          new Divider(),
          _buildItem("Custom Body Tip Dialog", () async {
            TipDialogHelper.show(new TipDialog.builder(
              bodyBuilder: (context) {
                return new Container(
                  width: 120.0,
                  height: 90.0,
                  alignment: Alignment.center,
                  child: new Text(
                    "Custom",
                    style: new TextStyle(color: Colors.white),

                    /// if TipDialogContainer are outside of MaterialApp,
                    /// here is a must to set
                    textDirection: TextDirection.ltr,
                  ),
                );
              },
              color: Colors.blue.withAlpha(150),
            ));
          }),
          new Divider(),
        ]));
  }
}
